### PR TITLE
#603 Root URL Configuration

### PR DESCRIPTION
**Pull request by ItsMeBrianD**
### Description

This pull request includes experimental support for `EVIDENCE_BASE_URL`; which enables hosting with Github pages.

There are a lot of moving pieces here, so this needs a pretty thorough run-through. 

### Checklist

- ~~[ ] For UI or styling changes, I have added a screenshot or gif showing before & after~~
- [x] I have added a [changeset](https://github.com/evidence-dev/evidence/blob/main/CONTRIBUTING.md#adding-a-changeset)

### CODE REVIEW

File: sites/example-project/src/components/ui/Sidebar.svelte [hunk 1]
```diff
@@ -9,12 +10,14 @@
 
 	// children of the index page
 	let firstLevelFiles = fileTree?.children;
+
+	$: console.log(fileTree.children);
```
Comment (Winterhart) [inline sites/example-project/src/components/ui/Sidebar.svelte:14]: A forgotten `console.log` ?

File: sites/example-project/src/components/ui/Sidebar.svelte [hunk 2]
```diff
@@ -38,7 +43,7 @@
 				<a
 					href="/settings"
 					class="settings-link"
-					class:selected={$page.url.pathname === '/settings'}
+					class:selected={$page.url.pathname === `$/settings`}
```
Comment (Winterhart) [inline sites/example-project/src/components/ui/Sidebar.svelte:46]: Are we sure we want that `$`?

File: sites/example-project/src/pages/text-and-metrics/markdown/+page.md [hunk 3]
```diff
@@ -1,3 +1,5 @@
+<script> import { base } from "$app/paths"; </script>
```
Comment (Winterhart) [inline sites/example-project/src/pages/text-and-metrics/markdown/+page.md:1]: So if I understand, if people wants `links` in their reports they will have to import `base`?

**Review by Winterhart — COMMENTED** [review]
Few questions.. wow it's a lot of changes!

File: sites/docs/docs/deployment/github-pages.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
+---
+    hide_table_of_contents: false
+    title: Github Pages
+---
+
+:::caution
+Deploying to Github pages requires more javascript knowledge than alternatives like [Evidence Cloud](/deployment/evidence-cloud) [Netlify](/deployment/netlify) or [Vercel](/deployment/vercel), if you are unsure, consider using another method.
```
Comment (archiewood) [inline sites/docs/docs/deployment/github-pages.md:7]: I would say "...requires some JavaScript, unlike alternatives, such as..." as many simple evidence apps need no js.

File: sites/docs/docs/deployment/github-pages.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+:::
+
+## Deploying to Github Pages
+
```
Comment (archiewood) [inline sites/docs/docs/deployment/github-pages.md:11]: Is it worth a short para on why?

GitHub pages are typically deployed with a base path that is not `/` eg ..., which is a non standard configuration, and requires some working around.

File: sites/docs/docs/deployment/github-pages.md [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+Deploying to Github Pages requires a few additional steps and considerations compared to other providers.
+
+First, [configure githug pages for your repository](https://docs.github.com/en/pages/quickstart).
+
+Once Github pages has been set up; build your project with the environment variable `EVIDENCE_BASE_PATH` set to your repositories name; note that it must start, but not end, with a `/`. This will configure evidence to point built-in links to `/your-base-path` instead of `/`. (`EVIDENCE_BASE_PATH={your repository} npm run build`)
+
+Finally, whenever you create a relative link (e.g. one that starts with `/` and navigates to a page in your evidence project), you must prefix it with `{base}`. This will ensure that the link functions properly when deployed to github pages.
```
Comment (archiewood) [inline sites/docs/docs/deployment/github-pages.md:18]: Is this the definition of a relative link? Sounds like an absolute link to me.

What about links like ../../my-cool-analysis/

File: packages/evidence/scripts/build-template.js [hunk 7]
```diff
@@ -50,6 +50,9 @@ fs.outputFileSync(
             files: {
                 routes: 'src/pages',
                 lib: 'src/components'
+            },
+            paths: {
+                base: process.env.EVIDENCE_BASE_PATH ?? undefined
```
[the remark this answers is not in the document]
Comment (ItsMeBrianD) [reply packages/evidence/scripts/build-template.js:55]: No if it is unset the behavior remains the same as before

File: sites/example-project/src/pages/text-and-metrics/markdown/+page.md (hunk 3 above)
Comment (ItsMeBrianD) [reply sites/example-project/src/pages/text-and-metrics/markdown/+page.md:1]: It was added to preprocess so it shouldn't be required, just needed in the example project

**Comment by mcrascal** [comment]
@ItsMeBrianD thanks -- are there any measures we can take to ensure we don't hard code relative urls into our components in the future, and break GH deployments?

**Review by Winterhart — CHANGES_REQUESTED** [review]
Code looks great, few minor `console.log` to remove.

 I've just tested in `localhost` I got one issue.
Here's how I can reproduce it

- I click on a normal page `Linear Chart`
- I click on `hierarchical Nav` on the side menu
- Then if you navigate back to `Linear Chart` it's not working

In the terminal:

```
3:03:12 PM [vite-plugin-svelte] /src/pages/hierarchical nav/+page.md:123:9 'base' is not defined
```

On the browser:

Screenshot 2023-04-27 at 3 08 00 PM

**Comment by archiewood** [comment]
This isn't so much a code review as more of a philosophical point

- This is a pretty far reaching change, and impacts how we'll develop components and features in the future.
- I worry it could be somewhat fragile, and increase the surface area of our product for edge cases.
- Github pages is a pretty common SSG, but it's by no means the only one, nor the most powerful.
  - Even when we do support it, the GH pages experience will be worse than others - you'll have to append the base path manually to each to get it to work.
  - Evidence becomes less "deploy anywhere" - the same source code will not do the same thing on all the platforms we support.

TLDR svelteKit decided not to support this out the box, and I wonder if we should be taking a similar approach.

I'd love to hear other opinions on this, but I wonder if this is a reasonable increase in complexity for not enough benefit

**Comment by ItsMeBrianD** [comment]
> @ItsMeBrianD thanks -- are there any measures we can take to ensure we don't hard code relative urls into our components in the future, and break GH deployments?

We can add an additional check to run a build with the `EVIDENCE_BASE_PATH` environment variable set; the build will likely break if the base path isn't respected.

> This is a pretty far reaching change, and impacts how we'll develop components and features in the future.

> I worry it could be somewhat fragile, and increase the surface area of our product for edge cases.

This definitely creates a fair amount of development overhead, so I could see this going either way

> Evidence becomes less "deploy anywhere" - the same source code will not do the same thing on all the platforms we support.

If you include `{base}`, and then build without the `EVIDENCE_BASE_PATH`, it will behave as if `{base]` isn't there; so in theory this shouldn't impact where you can deploy a project.

---

As to the overall viability of this feature; I think it really depends on how large the use-case / demand for Github Pages feature is. It does remain one of the easiest places to stick a demo / project.

**Comment by ItsMeBrianD** [comment]
@Winterhart I can't reproduce the issue that you are describing, did you rebuild the `preprocess` package after checking out the branch?

**Review by Winterhart — COMMENTED** [review]
I just re-run the branch and re-build all the components. 

The error seems gone 🥳  I get a `undefined` alert on the page is it normal?

CleanShot 2023-04-27 at 17 52 17

**Comment by mcrascal** [comment]
@ItsMeBrianD 

I would propose we put this on the shelf for now. After reviewing: 

1. Even in the nav and UI components here, we've got a handful of issues cropping up. All look like they'd be quick to fix, but this is such an annoying gotcha to leave behind for ourselves and contributors.  
2. It seems like this is breaking svelte kit's pre-load on-hover behaviour. 

In terms of actual user impact, I think this is pretty small. GH pages are a nice-to-have. 

Next step would just be to document the lack of GH pages support in a separate PR. 

We will likely need to tackle this for some trickier customer-embedded cases, so I doubt this will all be for nothing. I had originally preferred this solution over some sort of middleware / preprocessor approach, but I think when we do decide to revisit this, it'll be worth a more detailed look at one of those approaches.

**Comment by ItsMeBrianD** [comment]
@mcrascal Sounds good; I'll open another PR with the documentation for that.

File: sites/example-project/src/components/ui/Databases/DatabaseSettingsPanel.svelte [hunk 8]
```diff
@@ -43,7 +44,7 @@
 	async function runTest() {
 		const sleep = (ms) => new Promise((r) => setTimeout(r, ms));
 		await sleep(1000);
-		const res = await fetch('/api/testConnection.json', {
+		const res = await fetch(`${base}/api/testConnection.json`, {
```
Comment (jtshafer) [inline sites/example-project/src/components/ui/Databases/DatabaseSettingsPanel.svelte:47]: why not just use `./`
